File: Problems & Solutions/W/Wine Buying and Selling/24 December 2022/Java/Wine Buying and Selling.java
//{ Driver Code Starts
//Initial Template for Java

import java.io.*;
import java.util.*;

public class GFG {
    public static void main(String[] args) throws Exception {
        Scanner sc = new Scanner(System.in);
        int T = sc.nextInt();
        while(T>0)
        {
            int N = sc.nextInt();
            int arr[] = new int[N];
            for(int i=0; i<N; i++)
                arr[i] = sc.nextInt();
            
            Solution g = new Solution();
            long ans = g.wineSelling(arr,N);
            
            System.out.println(ans);
            T--;
        }
    }
}


// } Driver Code Ends

class Solution {
    static class pair{
        int first, second;
        pair(int first, int second){
            this.first = first;
            this.second = second;
        }
    }
    long wineSelling(int Arr[],int N){
        ArrayList<pair> buy = new ArrayList<>();
        ArrayList<pair> sell = new ArrayList<>();
        for(int i=0;i<N;i++){
            if(Arr[i]>0) 
                buy.add(new pair(Arr[i],i));
            else 
                sell.add(new pair(Arr[i],i));
        }
        long ans = 0;
        int i=0,j=0;
        while(i < buy.size() && j < sell.size()){
            long x = Math.min(buy.get(i).first,-sell.get(j).first);
            buy.get(i).first -= x;
            sell.get(j).first += x;
            long diff = Math.abs(buy.get(i).second - sell.get(j).second);
            ans += (x * diff);
            if(buy.get(i).first == 0) 
                i++;
            if(sell.get(j).first == 0) 
                j++;
        }
        return ans;
    }
}

//{ Driver Code Starts.
// } Driver Code Ends

File: Problems & Solutions/W/Wine Buying and Selling/24 December 2022/C++/Wine Buying and Selling.cpp
//{ Driver Code Starts
//Initial Template for C++

#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends


class Solution{
  public:
  long long int wineSelling(vector<int>& Arr, int N){
        vector<pair<int,int> > buy;
        vector<pair<int,int> > sell;
        for(int i=0;i<N;i++){
            if(Arr[i]>0) buy.push_back({Arr[i],i});
            else sell.push_back({Arr[i],i});
        }
        long long int ans = 0;
        int i=0,j=0;
        while(i<buy.size() and j<sell.size()){
            long long int x = min(buy[i].first,-sell[j].first);
            buy[i].first -= x;
            sell[j].first += x;
            long long int diff = abs(buy[i].second - sell[j].second);
            ans += (x * diff);
            if(buy[i].first == 0) i++;
            if(sell[j].first == 0) j++;
        }
        return ans;
  }
};



//{ Driver Code Starts.



int main(){
    int t;
    cin>>t;
    while(t--){
        int n;
        cin>>n;
        
        vector<int> arr(n);
        for(int i=0;i<n;i++){
            cin>>arr[i];
        }
        Solution ob;
        long long int ans = ob.wineSelling(arr,n);
        
        cout<<ans<<endl;
    }
}
// } Driver Code Ends